"yesterday" and single value support improvements for TimeIntervalArgs.

scott-gibson-sociomantic opened the pull request:
This PR adds the following features
1. specifying "yesterday" will use the unix timestamp of the previous day
2. specifying a single parameter now supported
    - With timestamp (including 'yesterday' it will be 24 hours range after given time.
    - With a duration provided it will be a range of X in the past to the current time.

it also adds an overload for the 'time' method to unsure the unit tests are using a stable value.

---

I've added a second commit to standardize how the interval is defined. It is to never include the final
parsed second. For example, `"-t 1h" with now:1000 (start:940, end:999)`. This was done after an internal poll and discussion.

scott-gibson-sociomantic committed:
Provide an overload for 'time(null)' in unit tests.

The existing unit tests were relying on real system timing
which can be prone to errors. This ensures the same value is always
used within the test suite.

scott-gibson-sociomantic committed:
Support single timestamp/interval arguments in TimeIntervalArgs.

The initial "default interval" didn't make practical sense when
using it in real applications. For useability purposes just specifying
a single date generally means you want to process the entire day. By
specifying a single time range you generally mean the last "x" days/hours
etc.

scott-gibson-sociomantic committed:
Support "yesterday" special case in TimeIntervalArgs.

scott-gibson-sociomantic commented:
One question that perhaps needs answering is if if the current second should be included in the 'end' of a range when calling with a single time interval. Eg. if i said create a range of 1d in the past, should it include the current second because when we provide a range of "2019-01-01 to 2019-02-01" we don't include 2019-02-01 00:00:00 in the range.

Edit:
Currently it is including the 'now' second, but i think it should be changed for consistency sake.

mihails-strasuns commented:
@scott-gibson-sociomantic I'd ask that question to actual users of this utility :) Neither seems more "correct" to me, it is simply a matter of practical convenience.

bogdan-szabo-sociomantic commented:
I think it's up to the app to use a closed or opened time interval. So probably there is enough to do `interval.end = interval.begin + SECONDS_IN_DAY;` here: https://github.com/sociomantic-tsunami/ocean/pull/691/files#diff-417c9e96a89af2c59b9e11a8237611adR117

otherwise LGTM

scott-gibson-sociomantic committed:
Standardize how parsing the end of a timerange is handled.

The parsed end second is never included in the resulting timerange.
This includes when using time ranges such as "100 1m" which will
now be 100->159 instead of 100->160

bogdan-szabo-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -2,6 +2,39 @@
+
+    Timerange
+        Gives X amount in the past NOT including the current second.
+        "-t 1h" with now:1000 (start:940, end:999)
```
with this case, `1h` = `59m59s`

I know that this is equivalent for `-t 3600s`, but this will just run for the last `3599` seconds, which might not make much sense :smile: 

I also wonder how this should work:
"-t 1s" with now:1000 (start:999, end:999)?

also, what `-t 0h` should do?

scott-gibson-sociomantic replied:
The '1h" was a mistake, it should say `1m`.
-t 3600s is still 3600 seconds worth of data, because it is intended that the ranges' end is to be used in the query eg.
```d
for ( long i = range.begin; i <= range.end; ++i )
```

`-t 1s` would mean you get the previous second of data. So you'd get the data from just 999.

-t 0h i believe should be an error (if it's not already)

scott-gibson-sociomantic replied:
Added a commit that treats 0 as an error.

bogdan-szabo-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -48,12 +82,17 @@ public void setupTimeIntervalArgs ( Arguments args, bool required )
             "an integer for unix timestamps\n\t" ~
             "a time duration `{int}m`. Supported units are [m]inutes, [h]ours and [d]ays\n\t" ~
             "an iso1806 date (YYYY-MM-DD)");
+    args("time-interval-exclude")
```
this argument might be too much ...

bogdan-szabo-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -76,10 +115,10 @@ public void setupTimeIntervalArgs ( Arguments args, bool required )
 {
-    if ( args["time-interval"].assigned.length != 2 &&
-        args["time-interval"].assigned.length != 0 )
+    if ( args["time-interval"].assigned.length < 1 ||
```
this might be more readable:
```
 auto num_args = args["time-interval"].assigned.length;
 if ( num_args < 1 || num_args > 2 )
```

scott-gibson-sociomantic replied:
Added

bogdan-szabo-sociomantic reviewed (approved):
Just a few notes, but it LGTM.

scott-gibson-sociomantic committed:
Treat time range of 0x as error in TimeIntervalArgs

bogdan-szabo-sociomantic commented:
can we merge this one? please?
